// dc-idom/index.js
// Generated by CoffeeScript 1.6.1
(function() {
  var $, IDom, debug, dispatch_impl, error, in_subtree, info, is_array, jqidom, warn, _ref,
    __slice = [].slice;

  dispatch_impl = require('libprotocol').dispatch_impl;

  _ref = dispatch_impl('ILogger', 'IDom'), info = _ref.info, warn = _ref.warn, error = _ref.error, debug = _ref.debug;

  is_array = require('libprotein').is_array;

  $ = require('commonjs-jquery');

  IDom = [['set-html!', ['new_content']], ['get-html', []], ['setValue', ['new_value']], ['setText', ['new_text']], ['getValue', []], ['alert', ['msg']], ['click', ['handler']], ['globalclick', ['handler']], ['keyDown', ['handler']], ['globalKeyDown', ['handler']], ['on_change', ['handler']], ['appendContent', ['content']], ['kill', []], ['stop_event', ['e']], ['setAttr', ['attr']], ['dbclick', ['e']], ['focusout', ['e']], ['focus', []], ['get_by_attr', ['attr']], ['get_by_id', ['id']], ['getData', ['attr', 'node']], ['get_id', ['node']], ['on_dom_ready', ['f']], ['one', ['sel']], ['document', []], ['get_root_node', []], ['add_event_listener', ['event_name', 'handler']], ['on_document_loaded', ['f']], ['toggleClass', ['from_to']], ['data', []], ['target', ['ev']], ['is_in', ['subtree', 'ev']], ['parent', []], ['disable', []], ['enable', []], ['text!', ['text']]];

  in_subtree = function($node, target) {
    if (!!$node.find(target).length || ($node.is(target))) {
      return true;
    } else {
      return false;
    }
  };

  jqidom = function(node) {
    var $node;
    $node = $(node);
    return {
      "text!": function(t) {
        return $node.text(t);
      },
      disable: function() {
        return $node.attr('disabled', 'disabled');
      },
      enable: function() {
        return $node.removeAttr('disabled');
      },
      data: function() {
        return $node.data();
      },
      parent: function() {
        return $node.parent();
      },
      target: function(ev) {
        return ev.target;
      },
      is_in: function(subtrees, target_node) {
        var elid, _i, _len;
        for (_i = 0, _len = subtrees.length; _i < _len; _i++) {
          elid = subtrees[_i];
          $node = elid === 'this' ? $node : $("#" + elid);
          if (in_subtree($node, target_node)) {
            return true;
          }
        }
        return false;
      },
      'set-html!': function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return $node.html(args.join(''));
      },
      'get-html': function() {
        return $node.html();
      },
      setValue: function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return $node.val(args.join(''));
      },
      setText: function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return $node.text(args.join(''));
      },
      getValue: function() {
        if (($node.prop('tagName')).toLowerCase() === 'span') {
          return $node.text();
        } else {
          return $node.val();
        }
      },
      setAttr: function(attr) {
        return say('setattr');
      },
      appendContent: function(content) {
        return $node.append("<div>" + content + "</div>");
      },
      alert: function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return alert.apply(null, args);
      },
      click: function(handler) {
        return $node.click(handler);
      },
      globalclick: function(handler) {
        return $(document).click(handler);
      },
      kill: function() {
        return $node.remove();
      },
      stop_event: function(e) {
        return $.Event(e).stopPropagation();
      },
      keyDown: function(handler) {
        return $node.bind('keydown', handler);
      },
      globalKeyDown: function(handler) {
        return $(document).bind('keydown', handler);
      },
      on_change: function(handler) {
        return $node.bind('onchange', handler);
      },
      dbclick: function(handler) {
        return $node.dblclick(handler);
      },
      focusout: function(handler) {
        return $node.blur(handler);
      },
      focus: function() {
        return $node.focus();
      },
      get_by_attr: function(attr) {
        var r1;
        r1 = (function() {
          try {
            if ($node.is(attr)) {
              return [$node];
            } else {
              return [];
            }
          } catch (e) {
            return [];
          }
        })();
        return r1.concat(($node.find(attr)).toArray());
      },
      get_by_id: function(id) {
        return $("#" + id);
      },
      getData: function(attr, node) {
        if (node == null) {
          node = $node;
        }
        return ($(node)).data()[attr];
      },
      get_id: function(node) {
        if (node == null) {
          node = $node;
        }
        return ($(node)).attr('id');
      },
      on_dom_ready: function(f) {
        return ($(document)).ready(f);
      },
      on_document_loaded: function(f) {
        return ($(window)).load(f);
      },
      one: function(sel) {
        return $(sel);
      },
      document: function() {
        return window.document;
      },
      get_root_node: function() {
        return node;
      },
      add_event_listener: function(event_name, handler) {
        if (node.addEventListener) {
          return node.addEventListener(event_name, handler);
        } else if (node.attachEvent) {
          return node.attachEvent("on" + event_name, handler);
        } else {
          error("Can't add event listener: no addEventListener nor attachEvent present", event_name, node);
          throw "Can't add event listener: no addEventListener nor attachEvent present";
        }
      },
      toggleClass: function(_arg) {
        var from, to;
        from = _arg[0], to = _arg[1];
        if ($node.hasClass(from)) {
          $node.removeClass(from);
          return $node.addClass(to);
        } else {
          $node.removeClass(to);
          return $node.addClass(from);
        }
      }
    };
  };

  module.exports = {
    protocols: {
      definitions: {
        IDom: IDom
      },
      implementations: {
        IDom: jqidom
      }
    }
  };

}).call(this);
